Extractcode libarchive system provided

aj4ayushjain opened the pull request:
This provides the support for the freebsd os in plugins which can be used for pip installation.

pombredanne commented:
@knobix you review would be great as @aj4ayushjain added support for FreeBSD there.

@aj4ayushjain are you sure that `lib_dir = 'lib'` is enough? I would have expected some absolute path of sorts on FreeBSD?

knobix commented:
> @aj4ayushjain are you sure that `lib_dir = 'lib'` is enough? I would have expected some absolute path of sorts on FreeBSD?

That's a good start but the `lib_dir` and the name of the `.so` need to be changed for FreeBSD. `libarchive` is shipped with FreeBSD and the location is `/usr/lib/libarchive.so` which is a symbolical link that points to the actual binary.

With FreeBSD 11.2- and 11.3-RELEASE `/usr/lib/libarchive.so` points to `/usr/lib/libarchive.so.6`. (= 3.3.3) 

With FreeBSD 12.0-RELEASE `/usr/lib/libarchive.so` points to `/usr/lib/libarchive.so.7`. (= 3.3.0)

With FreeBSD 13.0-CURRENT `/usr/lib/libarchive.so` points to `/usr/lib/libarchive.so.7` (= 3.4.0)

There's also an recent [package/port](https://www.freshports.org/archivers/libarchive/) of `libarchive` available. The default prefix for local packages is `/usr/local` so the constellation would be then:

`/usr/local/lib/libarchive.so` points to `/usr/local/lib/libarchive.so.13`. 

Thus it would be preferable to try to use the pkg/port version of `libarchive` in `/usr/local/lib/libarchive.so` first. If that version is not present then `libarchive` from the FreeBSD base system in `/usr/lib/libarchive.so` can be used as a fallback.

aj4ayushjain commented:
Thanks @knobix  for the review.
I just saw the paths from the https://www.freshports.org/archivers/libarchive/ page and was not sure of that.
What's the reason behind not to use the path to original binary `/usr/local/lib/libarchive.so.13`?

knobix commented:
> Thanks @knobix for the review.
> I just saw the paths from the https://www.freshports.org/archivers/libarchive/ page and was not sure of that.

Pleasure and thanks for the update! The given paths/files listed in the `pkg-plist` are relative to the installation prefix (usually `/usr/local`) which is the reason, why all those pathnames are somewhat shorter or missing the leading `/`.

> What's the reason behind not to use the path to original binary `/usr/local/lib/libarchive.so.13`?

I thought about it to reduce the need for maintenance. The version of the `.so` might be bumped in near future (e.g. `libarchive.so.14`) so one doesn't need to always change the source code to match the new name of the library.

aj4ayushjain commented:
> > What's the reason behind not to use the path to original binary `/usr/local/lib/libarchive.so.13`?
> 
> I thought about it to reduce the need for maintenance. The version of the `.so` might be bumped in near future (e.g. `libarchive.so.14`) so one doesn't need to always change the source code to match the new name of the library.

Thank you++ for the reply.
As i have used this path to original  elsewhere also so will update it soon.

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
+       
+	elif mainstream_system == 'freebsd' :
+            
+            if path.isdir('/usr/local/') :
```
Why a blank line?

aj4ayushjain replied:
I had thought this makes code much more readable that was the reason of inline comments and blank line spacing but will remove it.

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
-			'extractcode.libarchive.dll': join(lib_dir, 'libarchive.so.13'),
-		}
-		return locations
+                '''User defined flavor if installation directory differs'''
```
Why do you use triple quotes there?  a plain # would be better? also what do you mean exactly with this comment?

aj4ayushjain replied:
If the user has defined some other path of installation or some different flavor from unusual.

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
+            if path.isdir('/usr/local/') :
+
+                lib_dir = '/usr/local'
+
```
Same as above, this empty line is not needed there

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
+                lib_dir = '/usr/local'
+
+            else :
+
```
Same as above, this empty line is not needed there

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
-		
-			lib_dir = '/usr/lib64'
+            elif distribution in rpm_based_distro:
+
```
Why an empty line? IMHO not needed here

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
+        mainstream_system = platform.system()
+
+        if mainstream_system == 'Linux' :
+            
```
Empty line not needed here

pombredanne commented on plugins-builtin/extractcode-libarchive-system_provided/src/extractcode_libarchive/__init__.py:
```diff
@@ -28,36 +28,61 @@
-		debian_based_distro=['Ubuntu','Mint','debian']
+    def get_locations(self):
+
+        '''This defines the various mainstream system available 
```
We use triple double quotes """ for docstrings and no empty line before. We also use the imperative style and describe what is returned
So here this should be:
```
    def get_locations(self):
        """
        Return a mapping of {location key: location} providing the installation
        locations of the libarchive shared library as installed on various Linux
        distros or on FreeBSD.
        """
```

aj4ayushjain replied:
Will take care of these things from next time.
And will try to improve my imperative style comment.

aj4ayushjain replied:
Taken Note off!!
Will take care of these things from next time.
And will try to improve my imperative style comment.

aj4ayushjain commented:
@pombredanne  I have added the support for the freebsd in all the plugins and addressed the issue #1736.

pombredanne commented:
@aj4ayushjain re

>  I have added the support for the freebsd in all the plugins and addressed the issue #1736.

excellent! do you know what would be the procedure to test this for @madscientist159 ?

aj4ayushjain commented:
> @aj4ayushjain re
> 
> > I have added the support for the freebsd in all the plugins and addressed the issue #1736.
> 
> excellent! do you know what would be the procedure to test this for @madscientist159 ?

Thank you.
By test you meant how he can use it to solve the issue with binaries  or something else?

pombredanne commented:
@aj4ayushjain By test I meant some simple instructions here on how to test scancode on power pc.

pombredanne commented:
@aj4ayushjain I am merging this now anyway.

aj4ayushjain commented:
@pombredanne I have provided some steps related to POWER pc system in issue there which could be used to run scancode and i need feedback from them if problems if any and too merge is good enough.
